plugins/jobs/src/js/components/JobFormModal.tsx

```diff
@@ -102,39 +145,229 @@ export default class JobFormModal extends React.Component<
     isEdit: false,
     job: new Job()
   };
+  private triggerFormSubmit: FormSubmitCallback;
 
   constructor() {
-    super(...arguments);
+    super();
 
     this.state = {
-      defaultTab: ""
+      jobFormModel: null,
+      jobJsonString: null,
+      isJsonMode: false
     };
 
-    this.handleTabChange = this.handleTabChange.bind(this);
+    this.triggerFormSubmit = undefined;
+
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleJSONEditorChange = this.handleJSONEditorChange.bind(this);
+    this.handleFormChange = this.handleFormChange.bind(this);
+    this.handleTriggerSubmit = this.handleTriggerSubmit.bind(this);
+    this.handleInputModeToggle = this.handleInputModeToggle.bind(this);
   }
 
-  handleTabChange(tab: string) {
-    this.setState({ defaultTab: tab });
+  componentWillReceiveProps(nextProps: JobFormModalProps) {
+    if (nextProps.isOpen !== this.props.isOpen) {
+      this.resetState();
+    }
+  }
+
+  handleFormChange(event: { model: object }) {
+    if (!event.model) {
+      return;
+    }
+
+    this.props.handleErrorMessageChange(null);
+  }
+
+  handleSubmit() {
+    const { isEdit } = this.props;
+    const job = this.createJobFromEditorContents();
+    if (!job) {
+      return;
+    }
+
+    const jobSpec = JobUtil.createJobSpecFromJob(job);
+
+    this.props.handleSubmit(isEdit, jobSpec as JobData, job.getId());
+  }
+
+  handleInputModeToggle() {
+    this.setState({ isJsonMode: !this.state.isJsonMode });
+  }
+
+  handleJSONEditorChange(jobJsonString: string) {
+    this.setState({ jobJsonString }, () =>
+      this.props.handleErrorMessageChange(null)
+    );
+  }
+
+  handleTriggerSubmit(submitFunction: () => DataTriple) {
+    this.triggerFormSubmit = submitFunction;
+  }
+
+  resetState() {
+    const { job } = this.props;
+    this.setState({
+      jobFormModel: JobUtil.createFormModelFromSchema(JobSchema, job),
+      jobJsonString: JSON.stringify(
+        cleanJobJSON(JobUtil.createJobSpecFromJob(job as Job)),
+        null,
+        2
+      )
+    });
+  }
+
+  createJobFromEditorContents(keepValidationErrors = false) {
+    const { jobJsonString, isJsonMode } = this.state;
+    let jobDefinition;
+
+    // Try to parse JSON string and detect errors
+    try {
+      if (typeof jobJsonString !== "string") {
+        throw new Error();
+      }
+      jobDefinition = JSON.parse(jobJsonString);
+    } catch (e) {
+      this.props.handleErrorMessageChange({
+        message: "Invalid JSON syntax",
+        details: [{ path: "/", errors: [e.toString()] }]
+      });
+
+      return null;
+    }
+
+    const job = new Job(jobDefinition);
```

Comment:
Do we plan to move away from the job struct?

Reply:
Not in this PR I'd say

Reply:
Seems reasonable 👍